File: webapp/src/domain/calorieEstimator.ts
import type { Confidence, MealEstimate, MealItem } from "./models";
import { lookup } from "./nutritionDb";

/**
 * In-browser fallback estimator, a faithful port of the backend's
 * `estimator.py`: mock recognition + real reconciliation against the nutrition
 * table. Bands widen with uncertainty, and the meal's overall confidence is
 * the LOWEST item's — never an average. Results are flagged `fromFallback`.
 */

const BAND: Record<Confidence, number> = { HIGH: 0.1, MEDIUM: 0.2, LOW: 0.35 };
const ORDER: Record<Confidence, number> = { HIGH: 0, MEDIUM: 1, LOW: 2 };

type Recognized = [food: string, grams: number, confidence: Confidence];

const SAMPLES: Recognized[][] = [
  [
    ["grilled chicken thigh", 140, "HIGH"],
    ["white rice", 180, "LOW"],
    ["mixed salad", 90, "MEDIUM"],
  ],
  [
    ["margherita pizza", 220, "MEDIUM"],
    ["olives", 40, "HIGH"],
  ],
  [
    ["porridge", 300, "HIGH"],
    ["banana", 120, "HIGH"],
  ],
  [
    ["chicken curry", 250, "LOW"],
    ["naan", 90, "MEDIUM"],
    ["basmati rice", 150, "LOW"],
  ],
  [
    ["scrambled eggs", 150, "HIGH"],
    ["avocado toast", 130, "MEDIUM"],
  ],
  [
    ["beef burger", 250, "MEDIUM"],
    ["fries", 130, "LOW"],
  ],
];

const round1 = (v: number) => Math.round(v * 10) / 10;

function itemFor(food: string, grams: number, confidence: Confidence): MealItem {
  const row = lookup(food, grams);
  // An unmatched food can never be better than LOW confidence.
  const conf: Confidence = row.matched ? confidence : "LOW";
  const band = BAND[conf];
  return {
    name: row.food,
    grams,
    kcal: row.kcal,
    kcalLow: Math.round(row.kcal * (1 - band)),
    kcalHigh: Math.round(row.kcal * (1 + band)),
    protein: row.protein,
    carbs: row.carbs,
    fat: row.fat,
    confidence: conf,
  };
}

function fromItems(items: MealItem[], note: string): MealEstimate {
  const overall = items.reduce<Confidence>(
    (worst, i) => (ORDER[i.confidence] > ORDER[worst] ? i.confidence : worst),
    "HIGH",
  );
  const lows = items.filter((i) => i.confidence === "LOW");
  const lowest =
    lows.length > 0
      ? lows.reduce((min, i) => (i.kcal < min.kcal ? i : min)).name
      : null;
  return {
    items,
    totalKcal: items.reduce((s, i) => s + i.kcal, 0),
    totalLow: items.reduce((s, i) => s + i.kcalLow, 0),
    totalHigh: items.reduce((s, i) => s + i.kcalHigh, 0),
    protein: round1(items.reduce((s, i) => s + i.protein, 0)),
    carbs: round1(items.reduce((s, i) => s + i.carbs, 0)),
    fat: round1(items.reduce((s, i) => s + i.fat, 0)),
    overallConfidence: overall,
    lowestConfidenceItem: lowest,
    note,
    fromFallback: true,
    personaReaction: "",
  };
}

export function estimate(seed: number = Date.now(), note = ""): MealEstimate {
  const sample = SAMPLES[((seed % SAMPLES.length) + SAMPLES.length) % SAMPLES.length];
  return fromItems(
    sample.map(([food, grams, conf]) => itemFor(food, grams, conf)),
    note,
  );
}

/** Rebuild the meal after a one-tap portion correction on a single item. */
export function withPortion(
  est: MealEstimate,
  itemName: string,
  grams: number,
): MealEstimate {
  const items = est.items.map((item) => {
    if (item.name !== itemName) return item;
    // The user just told us the portion — that item is now HIGH confidence.
    return itemFor(item.name, grams, "HIGH");
  });
  return {
    ...fromItems(items, est.note),
    fromFallback: est.fromFallback,
    personaReaction: est.personaReaction,
  };
}

File: android/app/src/main/java/com/judgemycal/app/domain/CalorieEstimator.kt
package com.judgemycal.app.domain

import kotlin.math.roundToInt

/**
 * On-device fallback estimator, a faithful port of the backend's `estimator.py`:
 * mock recognition + real reconciliation against [NutritionDb]. Per-item bands
 * widen with uncertainty, and the meal's overall confidence is the LOWEST
 * item's — never an average — because one shaky portion makes the whole total
 * uncertain. Used only when the backend is unreachable (or no backend is
 * configured); results are flagged `fromFallback`.
 */
object CalorieEstimator {

    private val BAND = mapOf(
        Confidence.HIGH to 0.10,
        Confidence.MEDIUM to 0.20,
        Confidence.LOW to 0.35,
    )

    private data class Recognized(val food: String, val grams: Int, val confidence: Confidence)

    // Portion confidence reflects how hard the portion is to judge from a photo
    // (rice/curry are notoriously hard; a countable banana is easy).
    private val SAMPLES = listOf(
        listOf(
            Recognized("grilled chicken thigh", 140, Confidence.HIGH),
            Recognized("white rice", 180, Confidence.LOW),
            Recognized("mixed salad", 90, Confidence.MEDIUM),
        ),
        listOf(
            Recognized("margherita pizza", 220, Confidence.MEDIUM),
            Recognized("olives", 40, Confidence.HIGH),
        ),
        listOf(
            Recognized("porridge", 300, Confidence.HIGH),
            Recognized("banana", 120, Confidence.HIGH),
        ),
        listOf(
            Recognized("chicken curry", 250, Confidence.LOW),
            Recognized("naan", 90, Confidence.MEDIUM),
            Recognized("basmati rice", 150, Confidence.LOW),
        ),
        listOf(
            Recognized("scrambled eggs", 150, Confidence.HIGH),
            Recognized("avocado toast", 130, Confidence.MEDIUM),
        ),
        listOf(
            Recognized("beef burger", 250, Confidence.MEDIUM),
            Recognized("fries", 130, Confidence.LOW),
        ),
    )

    fun estimate(seed: Long = System.currentTimeMillis(), note: String = ""): MealEstimate {
        val sample = SAMPLES[((seed % SAMPLES.size + SAMPLES.size) % SAMPLES.size).toInt()]
        val items = sample.map { rec ->
            val row = NutritionDb.lookup(rec.food, rec.grams)
            // An unmatched food can never be better than LOW confidence.
            val conf = if (row.matched) rec.confidence else Confidence.LOW
            val band = BAND.getValue(conf)
            MealItem(
                name = row.food,
                grams = rec.grams,
                kcal = row.kcal,
                kcalLow = (row.kcal * (1 - band)).roundToInt(),
                kcalHigh = (row.kcal * (1 + band)).roundToInt(),
                protein = row.protein,
                carbs = row.carbs,
                fat = row.fat,
                confidence = conf,
            )
        }
        return fromItems(items, note)
    }

    /** Rebuild the meal after a one-tap portion correction on a single item. */
    fun withPortion(estimate: MealEstimate, itemName: String, grams: Int): MealEstimate {
        val items = estimate.items.map { item ->
            if (item.name != itemName) item else {
                val row = NutritionDb.lookup(item.name, grams)
                // The user just told us the portion — that item is now HIGH confidence.
                val band = BAND.getValue(Confidence.HIGH)
                item.copy(
                    grams = grams,
                    kcal = row.kcal,
                    kcalLow = (row.kcal * (1 - band)).roundToInt(),
                    kcalHigh = (row.kcal * (1 + band)).roundToInt(),
                    protein = row.protein,
                    carbs = row.carbs,
                    fat = row.fat,
                    confidence = Confidence.HIGH,
                )
            }
        }
        return fromItems(items, estimate.note).copy(
            fromFallback = estimate.fromFallback,
            personaReaction = estimate.personaReaction,
        )
    }

    private fun fromItems(items: List<MealItem>, note: String): MealEstimate {
        val overall = items.maxByOrNull { it.confidence.ordinal }?.confidence ?: Confidence.LOW
        val lows = items.filter { it.confidence == Confidence.LOW }
        return MealEstimate(
            items = items,
            totalKcal = items.sumOf { it.kcal },
            totalLow = items.sumOf { it.kcalLow },
            totalHigh = items.sumOf { it.kcalHigh },
            protein = round1(items.sumOf { it.protein }),
            carbs = round1(items.sumOf { it.carbs }),
            fat = round1(items.sumOf { it.fat }),
            overallConfidence = overall,
            lowestConfidenceItem = lows.minByOrNull { it.kcal }?.name,
            note = note,
            fromFallback = true,
        )
    }

    private fun round1(v: Double) = (v * 10).roundToInt() / 10.0
}
